cpp/oneapi/dal/table/backend/interop/host_homogen_table_adapter.hpp

```diff
@@ -0,0 +1,325 @@
+                            info.column_index,
+                            range{ info.row_begin_index, info.row_end_index });
+            }
+            catch (const bad_alloc&) {
+                return status_t(daal::services::ErrorMemoryAllocationFailed);
+            }
+            catch (const out_of_range&) {
+                return status_t(daal::services::ErrorIncorrectDataRange);
+            }
+            catch (const std::exception&) {
+                return status_t(daal::services::UnknownError);
+            }
+        }
+        return status_t();
+    }
+
+    bool check_row_indexes_in_range(const block_info& info) const {
+        const std::int64_t row_count = original_table_.get_row_count();
+        return info.row_begin_index < row_count && info.row_end_index <= row_count;
+    }
+
+    bool check_column_index_in_range(const block_info& info) const {
+        const std::int64_t column_count = original_table_.get_column_count();
+        return info.single_column_requested && info.column_index < column_count;
+    }
+
+    template <typename Accessor, typename BlockData, typename... Args>
+    void pull_values(block_desc_t<BlockData>& block,
+                     std::int64_t row_count,
+                     std::int64_t column_count,
+                     const Accessor& acc,
+                     array<BlockData>& values,
+                     Args&&... args) const {
+        // The following const_cast operation is safe only when this class is used for read-only
+        // operations. Use on write leads to undefined behaviour.
+
+        if (block.getBlockPtr() != acc.pull(values, std::forward<Args>(args)...)) {
+            auto raw_ptr = const_cast<BlockData*>(values.get_data());
+            auto data_shared =
+                daal::services::SharedPtr<BlockData>(raw_ptr, daal_object_owner(values));
+            block.setSharedPtr(data_shared, column_count, row_count);
+        }
+    }
+
+    host_homogen_table_adapter(const homogen_table& table, status_t& stat)
+            // The following const_cast is safe only when this class is used for read-only
+            // operations. Use on write leads to undefined behaviour.
+            : base(
+                  daal::data_management::DictionaryIface::equal,
+                  ptr_data_t{ const_cast<Data*>(table.get_data<Data>()), daal_object_owner(table) },
+                  table.get_column_count(),
+                  table.get_row_count(),
+                  stat),
+              is_rowmajor_(table.get_data_layout() == data_layout::row_major) {
+        if (stat.ok() == false) {
+            return;
+        }
+        else if (table.has_data() == false) {
+            stat.add(daal::services::ErrorIncorrectParameter);
+            return;
+        }
+
+        original_table_ = table;
+
+        this->_memStatus = daal::data_management::NumericTableIface::userAllocated;
+        this->_layout = daal::data_management::NumericTableIface::aos;
+    }
```

Comment:
I cannot find code that passes information about feature types (continuous/categorical/...) from the table to NumericTable, is this implemented?

Reply:
Thanks, I forget about that